File: nostos-core/src/main/java/sncf/oui/nostos/core/coordinates/Latitude.java
package sncf.oui.nostos.core.coordinates;

import org.immutables.value.Value.Check;
import org.immutables.value.Value.Immutable;
import org.immutables.value.Value.Parameter;

@Immutable
abstract class Latitude {

    private static final double MIN_LATITUDE = -90.;
    private static final double MAX_LATITUDE = 90.;

    static Latitude of(final double degrees) {
        return ImmutableLatitude.of(degrees);
    }

    @Parameter
    abstract double degrees();

    @Check
    void check() {
        if (degrees() <= MIN_LATITUDE || degrees() >= MAX_LATITUDE) {
            throw new IllegalArgumentException("latitude should be between -90° and 90°");
        }
    }
}

File: nostos-core/src/main/kotlin/io/fries/nostos/core/coordinates/Latitude.kt
package io.fries.nostos.core.coordinates

internal class Latitude(val degrees: Double) {

    init {
        require(degrees in MIN_DEGREES..MAX_DEGREES) { "Latitude degrees should be in the range [$MIN_DEGREES, $MAX_DEGREES]" }
    }

    companion object {
        private const val MIN_DEGREES = -90.0
        private const val MAX_DEGREES = 90.0

    }
}
